File: Baekjoon/9251_LCS/B_9251.py
first_word = input()
second_word = input()

max_num = 0
dp = [[0] * len(second_word) for _ in range(len(first_word))]
for i in range(len(first_word)):
    first_item = first_word[i]
    for j in range(len(second_word)):
        second_item = second_word[j]
        if first_item == second_item:
            if i == 0 or j == 0:
                dp[i][j] = 1
            else:
                dp[i][j] = dp[i - 1][j - 1] + 1
        else:
            dp[i][j] = max(dp[i][j - 1], dp[i - 1][j])

        if dp[i][j] > max_num:
            max_num = dp[i][j]

print(max_num)

File: Baekjoon/9251_LCS/B_9251.js
const fs = require('fs')
const input = fs.readFileSync('/dev/stdin').toString().trim().split('\n')

const firstWord = input[0].split('')
const secondWord = input[1].split('')

const firstWordLength = firstWord.length
const secondWordLength = secondWord.length

let wordArray = new Array(firstWordLength + 1).fill(0).map(v => new Array(secondWordLength + 1).fill(0))

for (let i = 0; i < firstWordLength; i++) {
    const target = firstWord[i]
    for (let j = 0; j < secondWordLength; j++) {
        const candidate = secondWord[j]
        if (target === candidate) {
            wordArray[i + 1][j + 1] = wordArray[i][j] + 1
        } else {
            wordArray[i + 1][j + 1] = Math.max(wordArray[i][j + 1], wordArray[i + 1][j])
        }
    }
}

console.log(wordArray[firstWordLength][secondWordLength])